pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,1015 @@
+	startTime := time.Now()
+	defer func() {
+		klog.V(2).InfoS("Finished syncing BGPPolicy", "BGPPolicy", bpName, "durationTime", time.Since(startTime))
+	}()
+
+	bp, err := c.bgpPolicyLister.Get(bpName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// If the BGPPolicy is deleted or not applied to the current Node anymore, do some cleanup.
+	if bp == nil || !c.matchedCurrentNode(bp) {
+		bpState, exists := c.getBGPPolicyState(bpName)
+		// If this is the effective BGPPolicy for the current Node, the BGPPolicy state should exist, and do some cleanup.
+		if exists {
+			// Stop the BGP server process on the current Node.
+			if err := bpState.bgpServer.Stop(c.ctx); err != nil {
+				return err
+			}
+			// Delete the BGPPolicy state.
+			c.deleteBGPPolicyState(bpName)
+		}
+
+		// Unbind the BGPPolicy from the current Node.
+		// If the BGPPolicy is the effective one for the current Node, try to pop a new effective BGPPolicy from the
+		// alternatives list to the queue. If the BGPPolicy is an alternative, remove it from the alternatives list.
+		if newEffectiveBP := c.unbindNodeFromBGPPolicy(bpName); newEffectiveBP != "" {
+			c.queue.Add(newEffectiveBP)
+		}
+		return nil
+	}
+
+	// Bind the BGPPolicy to the current Node. If the BGPPolicy is not as the effective one, just return.
+	if asEffective := c.bindNodeToBGPPolicy(bpName); !asEffective {
+		klog.InfoS("The Node has already got an effective BGPPolicy. This is as an alternative", "BGPPolicy", bpName)
+		return nil
+	}
+
+	// Retrieve the listen port, local AS number, and router ID from the current BGPPolicy to start the BGP server.
+	routerID, err := c.getRouterID()
+	if err != nil {
+		return err
+	}
+	listenPort := defaultBGPListenPort
+	if bp.Spec.ListenPort != nil {
+		listenPort = *bp.Spec.ListenPort
+	}
+	localASN := bp.Spec.LocalASN
+
+	var needUpdateBGPServer bool
+	// Get the BGPPolicy state.
+	bpState, exists := c.getBGPPolicyState(bpName)
+	if !exists {
+		// If the BGPPolicy state doesn't exist, meaning the BGPPolicy has not been enforced on the current Node, create
+		// state for the BGPPolicy, and then start the BGP server.
+		bpState = c.newBGPPolicyState(bp)
+		needUpdateBGPServer = true
+	} else {
+		// Check the listen port, local AS number and routerID. If any of them have changed, then start a new BGP server
+		// and stop the stale one.
+		needUpdateBGPServer = bpState.listenPort != listenPort ||
+			bpState.localASN != localASN ||
+			bpState.routerID != routerID
+	}
+
+	if needUpdateBGPServer {
+		// Start the new BGP server.
+		globalConfig := &bgp.GlobalConfig{
+			ASN:        uint32(localASN),
+			RouterID:   routerID,
+			ListenPort: listenPort,
+		}
+		bgpServer := newBGPServerFn(globalConfig)
+		if err := bgpServer.Start(c.ctx); err != nil {
+			return fmt.Errorf("failed to start BGP server: %w", err)
+		}
+
+		// Stop the stale BGP server if it exists.
+		if bpState.bgpServer != nil {
```

Comment:
Why to stop stale BGP server after new BGP server? Won't be port conflict if other configs changes but not port?

Reply:
Good question! I should stop the server first.